### DISCUSSION TITLE
How to stream a remote EPUB from an URL without downloading it?

### DISCUSSION THREAD

**Discussion by rhodeon** [readium/kotlin-toolkit#49] [Q&A] [answered]
I'm working on an app which needs to fetch an epub file from a remote server and display the content.
I've analyzed the remote fetching done in the Readium test app which downloads and adds the book to the app's root directory, but wanted to confirm if there's a way to directly stream and read the contents of the file on the server without having to save a copy of the complete book on the  user's device.
I've tried modifying the `url` parameter in the `BookshelfFragment` to test it out, but that just opens up a webview of the website which prompts me to download the file.

**Comment by mickael-menu** [comment, marked as answer]
EDIT: This will be possible in the upcoming Readium 3.0.

There's nothing available out of the box for this use case, but there are options:

* If the EPUB is served exploded on the server at a given URL, then you should be able to read it using an [`HttpFetcher`](https://github.com/readium/r2-shared-kotlin/blob/3e57eae357ead0def6a93dc758f5841e691798ce/r2-shared/src/main/java/org/readium/r2/shared/fetcher/HttpFetcher.kt#L27-L39). You will need to provide to `Streamer.open()` a custom implementation of [`PublicationAsset`](https://github.com/readium/r2-shared-kotlin/blob/3e57eae357ead0def6a93dc758f5841e691798ce/r2-shared/src/main/java/org/readium/r2/shared/publication/asset/PublicationAsset.kt#L16) which creates the `HttpFetcher`. Look at [`FileAsset`](https://github.com/readium/r2-shared-kotlin/blob/3e57eae357ead0def6a93dc758f5841e691798ce/r2-shared/src/main/java/org/readium/r2/shared/publication/asset/FileAsset.kt#L23) for an example implementation.
* If the EPUB is served as a single ZIP file on the server, it will be more complicated. The Java `ZipFile` API we are using doesn't support remote streaming of a ZIP archive. What you could do is:
    1. Find a ZIP library supporting streaming.
    2. Create an implementation of [`Archive`](https://github.com/readium/r2-shared-kotlin/blob/develop/r2-shared/src/main/java/org/readium/r2/shared/util/archive/Archive.kt) using this library. See for example [`JavaZip`](https://github.com/readium/r2-shared-kotlin/blob/develop/r2-shared/src/main/java/org/readium/r2/shared/util/archive/JavaZip.kt).
    3. Similarly to option 1, create a custom `PublicationAsset` but this time using an [`ArchiveFetcher`](https://github.com/readium/r2-shared-kotlin/blob/3e57eae357ead0def6a93dc758f5841e691798ce/r2-shared/src/main/java/org/readium/r2/shared/fetcher/ArchiveFetcher.kt#L28) and provide an instance of your custom `Archive` implementation.

**Reply by rhodeon** [reply]
Thanks for the response.
I tried out the first approach which seems promising, but I'm running into issues with it.

I created an implementation of `PublicationAsset` like you suggested:

```class UrlAsset(override val name: String) : PublicationAsset {
    override suspend fun createFetcher(
        dependencies: PublicationAsset.Dependencies,
        credentials: String?
    ): Try<Fetcher, Publication.OpeningException> {
        return try {
            val url = "https://github.com/rhodeon/readium_remote_test/tree/main/pg66407/"
            val httpClient = DefaultHttpClient()
            val fetcher = HttpFetcher(httpClient, url)
            Try.success(fetcher)
        } catch (e: Exception) {
            Try.failure(Publication.OpeningException.Forbidden(e))
        }
    }

    override suspend fun mediaType(): MediaType {
        return MediaType.EPUB
    }
}
```

I then intercepted the `openBook` function in the `BooksheltViewModel` as a test:

```
val asset = UrlAsset("pg66407")

streamer.open(asset, allowUserInteraction = true, sender = context)
    .onFailure {
        Timber.d(it)
        channel.send(Event.OpenBookError(it.getUserMessage(r2Application)))
    }
    .onSuccess {
        if (it.isRestricted) {
            it.protectionError?.let { error ->
                Timber.d(error)
                channel.send(Event.OpenBookError(error.getUserMessage(r2Application)))
            }
        } else {
            val url = prepareToServe(it)
            callback.invoke(book, asset, it, url)
        }
    }
```

Besides getting request errors for `rights.xml`, `encryption.xml` and `license.lcpl`, which are expected as they aren't in the exploded epub, I get an error due to a missing OPF file.
`2021-09-30 18:26:14.315 3699-3699/org.readium.r2reader D/BookshelfViewModel$openBook: org.readium.r2.shared.publication.Publication$OpeningException$ParsingFailed: java.lang.Exception: Unable to find an OPF file.`

This occurs even when the URL has a valid OPF file.
The URL I used for testing is: https://github.com/rhodeon/readium_remote_test/tree/main/pg66407

**Reply by mickael-menu** [reply]
You can't use this URL directly as it's not serving directly the files but GitHub's website. 

Maybe try enabling GitHub pages on your repo, this is what we do to serve this:

https://github.com/readium/webpub-manifest/tree/master/examples/MobyDick

as

https://readium.github.io/webpub-manifest/examples/MobyDick/

**Reply by rhodeon** [reply]
Okay, thanks a lot. I'll try that out.
